asvdb/asvdb.py

```diff
@@ -925,49 +938,191 @@ def __createLockfile(self, dirPath):
+                          "lock again...")
+                
+                time.sleep(1)
+                otherLockfileTimes = self.__updateS3LockfileTimes()
+
+            # All clear, create lock
+            if self.debugPrint:
+                print(f"All clear, setting lock {thisLockfile}")
+            self.s3Resource.Object(self.bucketName, thisLockfile).put()
+            
+            #Give S3 time to see the new lock
+            time.sleep(1)
+
+            # Check for a race condition where another lock could have been created
+            # while creating the lock for this instance.
+            otherLockfileTimes = self.__updateS3LockfileTimes()
+            
+            if otherLockfileTimes[1] != 0:
+                self.__releaseS3Lock()
+                randTime = (int(30 * random.random()) + 5) + random.random()
+                if self.debugPrint:
+                    print(f"Collision - waiting {randTime} seconds before "
+                          "trying to lock again.")
+                time.sleep(randTime)
+            else:
+                break
 
+            i += 1
 
-    def __releaseS3Lock(self, url):
-        # FIXME: write this
-        raise NotImplementedError
+            
+    def __updateS3LockfileTimes(self):
+        # Find lockfiles in S3 Bucket
+        response = self.s3Resource.Bucket(self.bucketName).objects \
+            .filter(Prefix=path.join(self.bucketKey, self.lockfilePrefix))
+
+        length = 0
+        for lockfile in response:
+            length += 1
+            if self.lockfileName in lockfile.key:
+                lockfile.delete()
+                length -= 1
+
+        return (response, length)
+
+
+    def __releaseS3Lock(self):
+        thisLockfile = path.join(self.bucketKey, self.lockfileName)
+        if self.debugPrint:
+            print(f"Removing lock {thisLockfile}")
+        self.s3Resource.Object(self.bucketName, thisLockfile).delete()
 
 
     ###########################################################################
     # S3 utilities
     ###########################################################################
-    def __downloadIfS3(self):
-        # self.localS3Copy = tempfile.TemporaryDirectory(...)
-        #
-        # Download bucket contents and unpack in self.localS3Copy.name
-        #
-        # Set all the internal locations to point to the downloaded files:
-        # self.confFilePath = path.join(self.localS3Copy.name, self.confFileName)
-        # self.resultsDirPath = path.join(self.localS3Copy.name, self.resultsDirName)
-        # self.benchmarksFilePath = path.join(self.resultsDirPath, self.benchmarksFileName)
-        return
+    def __downloadIfS3(self, bInfo=BenchmarkInfo(), results=False):
+        def downloadS3(bucket, ext):
+            bucket.download_file(
+                path.join(self.bucketKey, ext),
+                path.join(self.localS3Copy.name, ext)
+            )
+
+        if self.__isS3URL(self.dbDir):
```

Comment:
Minor suggestion, feel free to ignore if you want but you could get a level of indentation back if you did this instead:
```
if not(self.__isS3URL(self.dbDir)):
   return
```